File: examples/javascript/pagination.mjs
/**
 * Copyright 2026 Volstrata.com
 *
 * Cursor pagination, walked to completion against the capability catalog.
 * Docs: https://volstrata.com/docs/api-base-url
 *
 * Run:  node pagination.mjs [PAGE_SIZE]
 *       npm run pagination
 *
 * Pages `GET /api/v1/meta/capabilities?limit=25` — a Free, unauthenticated
 * capability that lists every published operation — until `next_cursor` comes
 * back null.
 *
 * One page is one request, so a full walk at limit=25 is around seven requests.
 * The loop is bounded by MAX_PAGES below: an unbounded cursor loop against a
 * live API is how a rate-limit budget disappears in a single run.
 */

import { VolstrataApiError, heading, paginate, show, summarize } from "./lib/volstrata.mjs";

/** Rows per page. Small on purpose here, so the walk actually has several pages. */
const LIMIT = Number(process.argv[2] ?? 25);

/** Hard stop. Reaching it means "stopped early", not "reached the end". */
const MAX_PAGES = 12;

heading("Cursor pagination — meta.capabilities");
show("path", "/api/v1/meta/capabilities");
show("page size", LIMIT);
show("page bound", MAX_PAGES);

/**
 * The paged envelope is uniform across the surface:
 *
 *   {"ok": true, "next_cursor": "<opaque>" | null, "limit": 25, "total": 175,
 *    "<rows>": [ … ]}
 *
 * `next_cursor` is the entire termination condition — null means this was the
 * last page. `total` is the matching row count across all pages, which is handy
 * for a progress line but is not a substitute for the null check.
 *
 * The rows themselves sit under the operation's own key, which differs per
 * operation, so this file finds the array rather than assuming a name. Printing
 * one page before writing the loop is the fastest way to learn the key you need.
 */
function rowsOf(page) {
  for (const [key, value] of Object.entries(page)) {
    if (Array.isArray(value)) return { key, rows: value };
  }
  return { key: null, rows: [] };
}

const rows = [];
let collectionKey = null;
let envelope = null;
let stoppedEarly = false;
let pages = 0;

try {
  // `paginate` is an async generator: it issues one request per iteration and
  // stops when `next_cursor` is null or the bound is reached. Cursors are opaque
  // — pass the string straight back, never parse, edit or construct one.
  for await (const { page, data, nextCursor } of paginate(
    "/api/v1/meta/capabilities",
    { limit: LIMIT },
    { maxPages: MAX_PAGES, pauseMs: 300 },
  )) {
    const found = rowsOf(data);
    collectionKey ??= found.key;
    envelope ??= data;
    rows.push(...found.rows);
    pages = page;

    console.log(
      `  page ${String(page).padStart(2)}  ` +
        `rows ${String(found.rows.length).padStart(3)}  ` +
        `running total ${String(rows.length).padStart(4)}  ` +
        `next_cursor ${nextCursor === null ? "null (last page)" : "present"}`,
    );

    // The bound, made visible: if the last page still had a cursor, there is
    // more data and this run simply chose not to ask for it.
    if (page === MAX_PAGES && nextCursor !== null) stoppedEarly = true;
  }
} catch (error) {
  heading("The walk stopped on an error");
  if (error instanceof VolstrataApiError) {
    console.log(error.describe("  "));
  } else {
    console.log(`  ${error?.message ?? error}`);
  }
  console.log("\n  Errors reference: https://volstrata.com/docs/api-errors");
  process.exitCode = 1;
}

if (envelope) {
  heading("Envelope fields (from the first page)");
  show("ok", envelope.ok);
  show("limit applied", envelope.limit ?? "(not reported)"); // read it; your requested size may be capped
  show("total", envelope.total ?? "(not reported)");
  show("rows key", collectionKey ?? "(none found)");
  if (envelope.limit === undefined || envelope.total === undefined) {
    console.log("  (not every paged capability reports limit and total — `next_cursor`");
    console.log("   is the only field the loop actually needs)");
  }

  heading("Result");
  show("pages fetched", pages);
  show("rows collected", rows.length);
  show("complete", stoppedEarly ? `no — stopped at the ${MAX_PAGES}-page bound` : "yes — next_cursor was null");

  if (rows.length > 0) {
    // Each capability row describes one operation: {name, method, path, params,
    // policy_key, tier, tier_name}. Display `tier_name` ("Free", "Edge") to a
    // human and branch on `tier` in software.
    heading("First rows");
    for (const row of rows.slice(0, 5)) {
      console.log(`  ${summarize(row, { depth: 1, maxKeys: 6 })}`);
    }

    const byTier = new Map();
    for (const row of rows) {
      const tier = row?.tier_name ?? row?.tier ?? "(unknown)";
      byTier.set(tier, (byTier.get(tier) ?? 0) + 1);
    }
    heading("Rows by plan floor");
    for (const [tier, count] of [...byTier].sort((a, b) => b[1] - a[1])) {
      show(tier, count);
    }
  }

  console.log("\n  This listing is the reason not to hard-code a capability list:");
  console.log("  it is free, unauthenticated, and always current.");
}

File: examples/python/pagination.py
"""Copyright 2026 Volstrata.com

Cursor pagination, done once, correctly.
Docs: https://volstrata.com/docs/api-base-url

    python pagination.py

One collection is paged to completion under a page bound. The endpoint used here
is the capability catalog itself (GET /api/v1/meta/capabilities), because it is
public, it is genuinely long, and paging it prints something you will want to
read anyway.

The contract, shared by every paged capability:

    GET /api/v1/<domain>/<name>?limit=25
        -> {"ok": true, "next_cursor": "<opaque>", ..., "<collection>": [...]}

    GET /api/v1/<domain>/<name>?limit=25&cursor=<opaque>
        -> the next page

    ... until "next_cursor": null

Three rules that are easy to get wrong:

  * A cursor is opaque. Pass the exact string back; never parse one, never build
    one, never assume it encodes an offset.
  * `next_cursor: null` is the only end-of-collection signal. An empty page is
    not a reliable one.
  * `limit` is a request, not a promise. The server clamps it to its own cap and
    the page you get back is whatever it decided to send.

17 of the published operations declare a cursor parameter. The rest return
everything they have in one response.
"""

import sys

from volstrata_helpers import (
    VolstrataError,
    heading,
    key_banner,
    paginate,
    row,
    show,
)

# Bound every loop that talks to a live API. 175 rows at 25 a page is 7
# requests; 12 leaves headroom if the catalog grows, and still guarantees this
# program stops.
MAX_PAGES = 12

PAGE_SIZE = 25


def main() -> int:
    print("VolStrata pagination -- https://volstrata.com")
    key_banner()

    heading("Paging GET /api/v1/meta/capabilities?limit={0}".format(PAGE_SIZE))

    rows = []
    pages = 0
    pending_cursor = None

    try:
        # paginate() is the generator in volstrata_helpers: it issues one GET per
        # page, follows next_cursor, and stops at the bound or at a null cursor.
        for page in paginate("/meta/capabilities", max_pages=MAX_PAGES, limit=PAGE_SIZE):
            pages += 1
            batch = page.get("capabilities") or []
            rows.extend(batch)
            pending_cursor = page.get("next_cursor")

            # The paged envelope is {ok, next_cursor, limit, total, <collection>}.
            # Not every capability populates all of the counters -- this one
            # reports `count` -- so read them defensively and rely on
            # next_cursor, which is always there.
            print("  page {0:>2}  rows={1:<3} next_cursor={2}".format(
                pages,
                len(batch),
                "null (last page)" if not pending_cursor else repr(pending_cursor),
            ))
            for counter in ("limit", "total", "count"):
                if counter in page and pages == 1:
                    row("  envelope " + counter, page[counter])
    except VolstrataError as exc:
        # A 400 `invalid_cursor` is what you get for a hand-edited cursor.
        print("  Paging stopped: {0}".format(exc))
        return 0

    heading("Result")
    row("pages fetched", pages)
    row("rows collected", len(rows))
    row("stopped because", "next_cursor was null" if not pending_cursor
        else "the {0}-page bound was reached".format(MAX_PAGES))
    if pending_cursor:
        print("  There is more to read. Resume from the cursor above rather than")
        print("  starting over -- that is what cursors are for.")

    # The catalog rows are the same shape the MCP tool list is projected from:
    # a dotted capability name, its HTTP method and path, its parameters, and
    # the plan floor as both a slug and a customer-facing name.
    if rows:
        heading("One row, in full")
        show(rows[0])

        heading("First few capabilities")
        for item in rows[:8]:
            print("  {0:<28} {1:<5} {2:<34} {3}".format(
                item.get("name", "?"),
                item.get("method", "?"),
                item.get("path", "?"),
                item.get("tier_name", "?"),
            ))

    print("\n  Same pattern, other collections: market.candles, news.feed,")
    print("  calendar.events, insider.filings, social.sentiment.")
    print("  Reference: https://volstrata.com/docs/api-base-url")
    return 0


if __name__ == "__main__":
    sys.exit(main())
